### PR TITLE
Cleaning

### PR DESCRIPTION
**Pull request by Zardoz89**
```
- No more warning about a false inlined function
- Added asserts
- Removed unnecessary memset.
- Initialization of readline_t with calloc
- Check malloc/calloc return
- Avoid possible free(NULL)
- Added some useful comments on test cases
- Clarify API on readme file and header comments
```

### CODE REVIEW

File: readline.c [hunk 1]
```diff
@@ -12,13 +13,18 @@
...
+  if (NULL == rl) {
+    return rl;
+  }
   size_t len = strlen(buffer);
 
-  rl->cursor = 0;
-  rl->line = 0;
   rl->buffer = (char*) malloc(len);
-  memset(rl->buffer, 0, len);
+  if( NULL == rl->buffer ) {
```
Comment (yorkie) [inline readline.c]: unnecessary blank at last

File: readline.c [hunk 2]
```diff
@@ -12,13 +13,18 @@
 
 readline_t *
 readline_new(char * buffer) {
-  readline_t * rl = (readline_t*) malloc(sizeof(readline_t));
+  readline_t * rl = (readline_t*) calloc(1, sizeof(readline_t));
+  if (NULL == rl) {
+    return rl;
```
Comment (yorkie) [inline readline.c]: How about using `assert` to make sure the object shouldn't be NULL?

File: readline.c [hunk 3]
```diff
@@ -30,19 +36,22 @@ readline_next(readline_t * rl) {
...
-    rl->buffer[cur++] != '\n' && 
+    rl->buffer[cur++] != '\n' &&
     cur <= buffer_len);
 
   len = cur - rl->cursor - 1;
   ret = (char*) malloc(len);
 
-  if (ret == NULL || (len == 0 && cur > buffer_len)) {
+  if (ret == NULL) {
+    return NULL;
```
Comment (yorkie) [inline readline.c]: ditto

File: readline.c [hunk 4]
```diff
@@ -30,19 +36,22 @@ readline_next(readline_t * rl) {
...
+  } else if (len == 0 && cur > buffer_len) {
     free(ret);
     return NULL;
   }
 
-  memset(ret, 0, len);
```
Comment (yorkie) [inline readline.c]: Why did you remove this call? it actually reset all char code to ZERO.

File: readline.c [hunk 5]
```diff
@@ -67,11 +78,12 @@ readline_last_from_rl(readline_t * rl) {
   ret = (char*) malloc(len);
 
   if (ret == NULL) {
+    return NULL;
```
Comment (yorkie) [inline readline.c]: ditto, use `assert`?

**Comment by yorkie** [comment]
Thank you @Zardoz89 for the work, but actually this module seems to be deprecated in my side, as tj told me at [here](https://github.com/clibs/clib/issues/21#issuecomment-33294863):

> getline() / getdelim() might be similar (haven't checkout the code yet)

File: readline.c (hunk 4 above)
Comment (Zardoz89) [inline readline.c]: memcpy is already overwriting all the buffer. I don't see why is necessary to fill it with 0's when you are going to overwrite it.

File: readline.c (hunk 5 above)
Comment (Zardoz89) [inline readline.c]: Assert's aren't for checking errors. Assert's indicates assumption of a code. On other words, an assertion specifies that a program/library satisfies certain conditions at particular points in its execution.
Assert macro are only active when the source code is compiled with debug flags. Release code, have assert macro disabled. So when the library is compiled on debug mode, the user or the developer of the lib, could see if a problem comes of a bad usaged of the lib or of a real bug on the lib. For example, if a user code pass a bad readline_t where buffer is NULL, the user code is wrong, and the debug version of the lib would fail showing that.

More info : http://ptolemy.eecs.berkeley.edu/~johnr/tutorials/assertions.html

**Comment by Zardoz89** [comment]
Well, I did this becasue someone is trying to add it to FUZIX : https://github.com/EtchedPixels/FUZIX

**Comment by yorkie** [comment]
Okay, merged, I have not too much skills on clang, you are right.
